src/solvers/dg/dg_2d.jl

```diff
@@ -0,0 +1,594 @@
+  # cache::Dict{Symbol, Any}
+  # thread_cache::Any # to make fully-typed output more readable
+
+  # Create the basic cache
+  # Get cells for which an element needs to be created (i.e. all leaf cells)
+  leaf_cell_ids = leaf_cells(mesh.tree)
+
+  # TODO: Taal refactor, we should pass the basis as argument,
+  # not polydeg, to all of the following initialization methods
+  elements = init_elements(leaf_cell_ids, mesh,
+                           RealT, nvariables(equations), polydeg(dg))
+
+  interfaces = init_interfaces(leaf_cell_ids, mesh, elements,
+                               RealT, nvariables(equations), polydeg(dg))
+
+  # TODO: Taal implement BCs
+  boundaries, n_boundaries_per_direction = init_boundaries(leaf_cell_ids, mesh, elements,
+                               RealT, nvariables(equations), polydeg(dg))
+
+  mortars = init_mortars(leaf_cell_ids, mesh, elements,
+                         RealT, nvariables(equations), polydeg(dg), dg.mortar)
+
+  cache = (; elements, interfaces, boundaries, mortars)
+
+  # TODO: Taal refactor
+  # For me,
+  # - surface_ids, cell_ids in elements
+  # - neighbor_ids, orientations in interfaces
+  # - neighbor_ids, orientations, neighbor_sides in boundaries
+  # - neighbor_ids, large_sides, orientations in mortars
+  # seem to be important information about the mesh.
+  # Shall we store them there?
+
+  # Add specialized parts of the cache required to compute the volume integral etc.
+  cache = (;cache..., create_cache(mesh, equations, dg.volume_integral)...)
+  cache = (;cache..., create_cache(mesh, equations, dg.mortar)...)
+
+  return cache
+end
+
+
+function create_cache(mesh::TreeMesh{2}, equations, volume_integral::VolumeIntegralFluxDifferencing)
+  create_cache(mesh, have_nonconservative_terms(equations), equations, volume_integral)
+end
+
+function create_cache(mesh::TreeMesh{2}, nonconservative_terms::Val{false}, equations, ::VolumeIntegralFluxDifferencing)
+  NamedTuple()
+end
+
+# function create_cache(mesh::TreeMesh{2}, nonconservative_terms::Val{true}, equations, ::VolumeIntegralFluxDifferencing)
+#   # TODO: Taal implement if necessary
+#   NamedTuple()
+# end
+
+# function create_cache(mesh::TreeMesh{2}, equations, ::VolumeIntegralShockCapturingHG)
+#   # TODO: Taal implement
+# end
+
+function create_cache(mesh::TreeMesh{2}, equations, mortar_l2::LobattoLegendreMortarL2)
+  # TODO: Taal compare performance of different types
+  MA2d = MArray{Tuple{nvariables(equations), nnodes(mortar_l2)}, real(mortar_l2)}
+  # A2d  = Array{real(mortar_l2), 2}
+
+  fstar_upper_threaded = [MA2d(undef) for _ in 1:Threads.nthreads()]
+  fstar_lower_threaded = [MA2d(undef) for _ in 1:Threads.nthreads()]
+
+  (; fstar_upper_threaded, fstar_lower_threaded)
+end
+
+
+function wrap_array(u::AbstractVector, mesh::TreeMesh{2}, equations, dg::DG, cache)
+  unsafe_wrap(Array{eltype(u), ndims(mesh)+2}, pointer(u),
+              (nvariables(equations), nnodes(dg), nnodes(dg), nelements(dg, cache)))
+end
```

Comment:
What do we need this for? `wrap_array` seems to pop up in a lot of different places and even after reading the docs for `unsafe_wrap` I am not sure why this is necessary...

Reply:
That's because of AMR, see https://github.com/trixi-framework/Trixi.jl/pull/187/files#diff-12ba8aaaaed9876320167642b01d7dffR265-R273

Reply:
I haven't implemented AMR yet, so I'm not sure whether that solution works.

Reply:
OK, but at least it seems like https://github.com/SciML/OrdinaryDiffEq.jl/pull/1275 was already merged. Does that mean that we'll use `ElasticArrays.jl` then?

Reply:
No, I want to try out my wrapping and using plain `Vector`s on the DiffEq side, basically because of
> Needing to use ElasticArray doesn't feel completely good to me, since we also want to experiment with other array types such as PaddedMatrices.jl, see trixi-framework/Trixi.jl#166. Then, we would need to wrap an Array inside something from PaddedMatrices.jl inside something from ElasticArrays.jl - or the other way round? Is that possible at all? If we go further, this looks like it could easily explode.

Reply:
OK, I understand. Maybe it would be good to then call `u` something else here, wouldn't it? We use it extensively in DG, and anyone reading the code there will assume that `u` is the solver's `u`. I saw you sometimes called the DiffEq solution `u_vector` (which I guess is strictly only correct for the current way to implement it), sometimes you call it just `u` and the "normal" solution `u_wrapped` (e.g., https://github.com/trixi-framework/Trixi.jl/pull/187/files#diff-b67f54c56aa8e15588adb3c7f3512a18R75-R78).

I think it would be good to be consistent about this to avoid confusion. Maybe always call the DiffEq `u` something like `u_ode` and "regular" just `u`? I am not set on these particular names, but I think consistency would be helpful.

Reply:
Should be done once everything works :+1:

Reply:
I'll add it to the list...